### PR TITLE
Fix the regression of #1360: Impossible to delete a non-existent source

### PR DESCRIPTION
**Pull request by martinvlk**
Right now this pull request contains a partial fix, the test I added is still not passing.
I am going public with it now so that I can discuss some specifics with the others - details to be added later.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
Can you please look into the Travis failure?

**Comment by 23Skidoo** [comment]
Ah, OK, I see that you're aware of the failure.

File: cabal-install/Distribution/Client/IndexUtils.hs [hunk 1]
```diff
@@ -307,12 +307,15 @@ extractPkg entry blockNo = case Tar.entryContent entry of
   Tar.OtherEntryType typeCode content _
     | Tar.isBuildTreeRefTypeCode typeCode ->
       Just $ do
-        let path   = byteStringToFilePath content
-            err = "Error reading package index."
-        cabalFile <- tryFindAddSourcePackageDesc path err
-        descr     <- PackageDesc.Parse.readPackageDescription normal cabalFile
-        return $ BuildTreeRef (refTypeFromTypeCode typeCode) (packageId descr)
-                              descr path blockNo
+        let path = byteStringToFilePath content
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/IndexUtils.hs]: So the problem is that `extractPkg` is called from `parsePackageIndex`, which is called from `updatePackageIndexCacheFile`, which is called from `removeBuildTreeRefs`. But isn't `updatePackageIndexCacheFile` called only after `doRemove` - that is, after the invalid refs were already removed? So why should we fix `extractPkg` then - maybe it just suffices to improve the error message?

**Comment by martinvlk** [comment]
I did more investigation yesterday and I found out that yes, it may be a documentation problem, at least in part, but for a different reason than you suggest.

The thing is that when the directory doesn't exist, then the reference is NOT removed from the index, unless you specify its full path, as recorded in the index. This is because canonicalization of a nonexistent directory in `removeBuildTreeRefs` fails and therefore you never get to reconstruct the full path from a relative one specified on the command line. Also in this scenario the `updatePackageIndexCacheFile` bails out with a "does not exist" for the file and wipes the whole cache file, instead of just removing the bad reference.

The former of the two issues is hard to fix, and the only thing we can possibly do is document this in the user guide. The user has to specify the full path as given by `cabal sandbox list-sources` and then it gets removed successfully.

The problem of wiping the index cache file then is fixed by my commit and I think we should merge it, along with a documentation update. If you agree, I'll put together the doc update and also add tests to capture the behaviour.

**Comment by 23Skidoo** [comment]
We can update the manual, but it's more important to make `delete-source` print a helpful warning when it's can't canonicalise the provided path. Something like:

```
$ cabal sandbox delete-source foo
Can't figure out the absolute path 'foo' refers to. If you are trying to delete a reference to a removed directory, please provide the full absolute path (as given by `sandbox list-sources`).
```

Also, `remove-source` should fail even before it gets to updating the cache in this case.

> The latter problem of wiping the index cache file then is fixed by my commit and I think we should merge it

Yes, since `updatePackageIndexCacheFile` is also called from other places.

**Comment by martinvlk** [comment]
Ok, this is now ready for review.

I have streamlined the error handling as we are now dealing with different types of errors within a possibly list of input values. Each type of error will be reported only once for all affected paths and the same goes for success.

**Comment by 23Skidoo** [comment]
Can you please rebase this against master to get rid of the spurious merge commits? Also would be nice to see some of the commits squashed together to provide a more logical view of the changes.

File: cabal-install/Distribution/Client/Sandbox.hs [hunk 2]
```diff
@@ -452,26 +455,47 @@ sandboxDeleteSource verbosity buildTreeRefs _sandboxFlags globalFlags = do
...
   notice verbosity $ "Note: 'sandbox delete-source' only unregisters the " ++
     "source dependency, but does not remove the package " ++
     "from the sandbox package DB.\n\n" ++
     "Use 'sandbox hc-pkg -- unregister' to do that."
   where
-    convertWith dict pth = fromMaybe pth $ fmap fst $ find ((==pth) . snd) dict
-    showL = concat . intersperse " " . fmap show
+    showPaths f = concat . intersperse " " . fmap (show . f)
+
+    putNewlineStdErr = hPutStr stderr "\n"
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Sandbox.hs]: Why is this needed? Can't you just use `warn`?

File: cabal-install/Distribution/Client/Sandbox.hs [hunk 3]
```diff
@@ -452,26 +455,47 @@ sandboxDeleteSource verbosity buildTreeRefs _sandboxFlags globalFlags = do
...
+    showE [] = return ' '
+    showE errs@(Index.ErrNonregisteredSource{}:_) = showPaths Index.nrPath errs
+    showE errs@(Index.ErrNonexistentSource{}:_) = showPaths Index.nePath errs
+
+    errorType Index.ErrNonregisteredSource{} Index.ErrNonregisteredSource{} = True
+    errorType Index.ErrNonexistentSource{} Index.ErrNonexistentSource{} = True
+    errorType _ _ = False
+
+    handleErrors [] = return ()
+    handleErrors errs@(Index.ErrNonregisteredSource{}:_) = putNewlineStdErr >> warn verbosity ("Sources not registered: " ++ showE errs)
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Sandbox.hs]: Please try to stay within 80-col limit.

File: cabal-install/Distribution/Client/Sandbox/Index.hs [hunk 4]
```diff
@@ -156,39 +158,56 @@ addBuildTreeRefs verbosity path l' refType = do
...
+                       | ErrNonexistentSource   { nePath :: FilePath } deriving Show
+
 -- | Remove given local build tree references from the index.
 --
 -- Returns a tuple consisting of removed build tree refs and mappings from
 -- provided build tree refs to corresponding full directory paths).
 removeBuildTreeRefs :: Verbosity -> FilePath -> [FilePath]
-                       -> IO ([FilePath], [(FilePath, FilePath)])
+                       -> IO ([Either DeleteSourceError FilePath],
+                              (FilePath -> FilePath)) -- ^ A tuple consisting of:
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Sandbox/Index.hs]: I changed the Haddock string for this function, so this comment is no longer needed.

**Comment by 23Skidoo** [comment]
Looks OK, but please rebase your branch and squash related commits together.

File: cabal-install/Distribution/Client/Sandbox.hs (hunk 2 above)
Comment (martinvlk) [inline cabal-install/Distribution/Client/Sandbox.hs]: It formats lists of paths for inclusion in messages which then get reported by functions like `warn` or `notice`. It is needed, unless there is a more appropriate way to do it I am not aware of.

Comment (23Skidoo) [inline cabal-install/Distribution/Client/Sandbox.hs]: I was talking about `putNewlineStderr`.

**Comment by martinvlk** [comment]
I'm struggling a little with git here, squashing commits, etc. Hope the result will be what we expect. Also after squashing, you seem to be recorded as an author of my changes on the commit... oh well.

File: cabal-install/Distribution/Client/Sandbox.hs (hunk 2 above)
Comment (martinvlk) [inline cabal-install/Distribution/Client/Sandbox.hs]: `putNewlineStderr` just adds a newline to stderr, whereas `warn` will also prepend the string "Warning: ", which is undesirable for my purposes.

**Comment by 23Skidoo** [comment]
That's probably because you squashed some of your commits with mine. Try doing a rebase first, and then squash. You can go back to a previous state with `git reflog` and `git reset`.

File: cabal-install/Distribution/Client/Sandbox.hs (hunk 2 above)
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Sandbox.hs]: But why do you need a newline before `Warning: ...`? We don't do that in other places in the code. Can you show an example of output with and without the newline?

**Comment by martinvlk** [comment]
Closing this pull request, will create a new one from a cleaned-up branch.
